File: solutions/337/main.py
from typing import List 

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def robTree(self, node):
        result = [0, 0]
        if node is None: 
            return result
        left = self.robTree(node.left)
        right = self.robTree(node.right)
        # do not steal from current node 
        result[0] = max(left[0], left[1]) + max(right[0], right[1])
        # steal from current node 
        result[1] = node.val + left[0] + right[0]
        return result 
    def rob(self, root: TreeNode) -> int:
        result = self.robTree(root) 
        return max(result[0], result[1])

if __name__ == "__main__": 
    root = TreeNode(3)
    node1 = TreeNode(2)
    node2 = TreeNode(3)
    root.left = node1 
    root.right = node2 
    node3 = TreeNode(3)
    node1.right = node3
    node4 = TreeNode(1)
    node2.right = node4
    
    s = Solution()
    assert s.rob(root) == 7

File: solutions/337/main.rs
Struct Solution;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn rob(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut dp = vec![vec![0; 2]; 10000];
        let mut ans = 0; 

        fn traverse(root: Option<Rc<RefCell<TreeNode>>>) -> (i32, i32) {
            let mut result = (0, 0); 

            if root.is_none() {
                return result; 
            }

            let root = root.unwrap(); 

            let left = traverse(root.borrow().left.clone()); 
            let right = traverse(root.borrow().right.clone()); 

            // do not steal from cur node 
            result.0 = left.0.max(left.1) + right.0.max(right.1);
            // steal from cur 
            result.1 = root.borrow().val + left.0 + right.0;

            return result;  
        }
        
        let result = traverse(root);

        result.0.max(result.1) 
    }
}